pkg/api/routes.go

```diff
@@ -1642,6 +1677,84 @@ func (rh *RouteHandler) ListExtensions(w http.ResponseWriter, r *http.Request) {
 	WriteJSON(w, http.StatusOK, extensionList)
 }
 
+// github Oauth2 CodeExchange callback.
+func (rh *RouteHandler) GithubCodeExchangeCallback() rp.CodeExchangeCallback {
+	return func(w http.ResponseWriter, r *http.Request,
+		tokens *oidc.Tokens, state string, relyingParty rp.RelyingParty,
+	) {
+		ctx := r.Context()
+
+		client := github.NewClient(relyingParty.OAuthConfig().Client(ctx, tokens.Token))
+
+		email, groups, err := GetGithubUserInfo(ctx, client, rh.c.Log)
+		if email == "" || err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+
+			return
+		}
+
+		callbackUI, err := OAuth2Callback(rh.c, w, r, state, email, groups)
+		if err != nil {
+			if errors.Is(err, zerr.ErrInvalidStateCookie) {
+				w.WriteHeader(http.StatusUnauthorized)
+			}
+
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+
+		if callbackUI != "" {
+			http.Redirect(w, r, callbackUI, http.StatusFound)
+
+			return
+		}
+
+		w.WriteHeader(http.StatusCreated)
+	}
+}
+
+// Openid CodeExchange callback.
+func (rh *RouteHandler) OpenIDCodeExchangeCallback() rp.CodeExchangeUserinfoCallback {
+	return func(w http.ResponseWriter, r *http.Request, tokens *oidc.Tokens, state string,
+		relyingParty rp.RelyingParty, info oidc.UserInfo,
+	) {
+		email := info.GetEmail()
+		if email == "" {
+			rh.c.Log.Error().Msg("couldn't set user record for empty email value")
+			w.WriteHeader(http.StatusUnauthorized)
+
+			return
+		}
+
+		var groups []string
+
+		val, ok := info.GetClaim("groups").([]interface{})
+		if !ok {
+			rh.c.Log.Info().Msgf("couldn't find any 'groups' claim for user %s", email)
+		}
+
+		for _, group := range val {
+			groups = append(groups, fmt.Sprint(group))
+		}
+
+		callbackUI, err := OAuth2Callback(rh.c, w, r, state, email, groups)
```

Comment:
every time a user logs in with openid/ldap we save its groups in DB. (we don't skip getting his groups because that may change between autentications.)
every time a user logs in with session cookie we only read from from DB.